timelock-impl/src/test/java/com/palantir/atlasdb/timelock/adjudicate/LeaderElectionDurationAccumulatorTest.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.adjudicate;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+
+import com.google.common.util.concurrent.Uninterruptibles;
+import com.palantir.common.concurrent.PTExecutors;
+import com.palantir.conjure.java.lib.SafeLong;
+import com.palantir.timelock.feedback.LeaderElectionDuration;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.LongConsumer;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.LongStream;
+import org.junit.Test;
+
+public class LeaderElectionDurationAccumulatorTest {
+    private static final UUID LEADER_1 = UUID.randomUUID();
+    private static final UUID LEADER_2 = UUID.randomUUID();
+    private static final UUID LEADER_3 = UUID.randomUUID();
+
+    private LongConsumer mockConsumer = mock(LongConsumer.class);
+    private LeaderElectionDurationAccumulator accumulator = new LeaderElectionDurationAccumulator(mockConsumer, 5);
+
+    @Test
+    public void nothingConsumedWithFewUpdates() {
+        leaderElectionResultsWithDurationInRandomOrder(LEADER_1, LEADER_2, 4, 1);
+        leaderElectionResultsWithDurationInRandomOrder(LEADER_2, LEADER_3, 4, 5);
+
+        verifyNoInteractions(mockConsumer);
+    }
+
+    @Test
+    public void consumeAfterFiveUpdatesWithMinDuration() {
+        leaderElectionResultsWithDurationInRandomOrder(LEADER_1, LEADER_2, 5, 1);
+
+        verify(mockConsumer).accept(1L);
+        verifyNoMoreInteractions(mockConsumer);
+    }
+
+    @Test
+    public void consumeOnlyOnceUsingMinFromFirstFiveUpdates() {
+        leaderElectionResultsWithDurationInRandomOrder(LEADER_1, LEADER_2, 5, 15);
+        leaderElectionResultsWithDurationInRandomOrder(LEADER_1, LEADER_2, 10, 1);
+
+        verify(mockConsumer).accept(15L);
+        verifyNoMoreInteractions(mockConsumer);
+    }
+
+    @Test
+    public void consumeOnceForEachLeaderPair() {
```

Comment:
I think it would be best if the range of values for each pair (other than the repeated ones) should be non-intersecting - i.e. for leader (2,3) you should set the min to say 200 to make 100% certain that you aren't getting a spurious result.

Reply:
Only L1->L3 _should_ have generated the 7, but if you have a min of 5 and 10 requests, then there exists a request of time 7 there, even if they should report 5. By separating this, you _guarantee_ that it is coming from the right place.

Reply:
discussed offline: subject to the one-from-one-election assumption, that should be fine as if, e.g. the [7,11] range gave a 10 instead, there's no way for the [10,14] or [15,19] ranges to push a 7. Though let's just discard the need for that assumption and use disjoint ranges.